Add DTO to prevent over-posting on API (#17074)
dotnet/AspNetCore.Docs: aspnetcore/tutorials/first-web-api.md

* Set the URI of the object to delete (for example `https://localhost:5001/api/TodoItems/1`).
* Select **Send**.

<a name="over-post"></a>

## Prevent over-posting

> Currently the app exposes the entire `TodoItem`. Productions apps typically limit the data that's input and returned using subset of the model. The subset model is called a Data Transfer Object (DTO), input model, or view model.

Reviewer: What about trying to rephrase this first paragraph as follows? Feel free to dismiss this suggestion, if you don't like it.
```suggestion
Currently our sample app exposes the entire `TodoItem` object. Productions apps typically limit the data that's input and returned using subset of the model. There are multiple reasons behind this and security aspect is a major one. That subset of a model is usually referred to as a Data Transfer Object (DTO), input model, or view model. We will use the `DTO` naming in this article.
```

> Currently the sample app exposes the entire `TodoItem` object. Productions apps typically limit the data that's input and returned using a subset of the model. There are multiple reasons behind this and security aspect is a major one. The subset of a model is usually referred to as a Data Transfer Object (DTO), input model, or view model. **DTO** is used in this article.

Reviewer 2: ```suggestion
Currently the sample app exposes the entire `TodoItem` object. Productions apps typically limit the data that's input and returned using a subset of the model. There are multiple reasons behind this and security is a major one. The subset of a model is usually referred to as a Data Transfer Object (DTO), input model, or view model. **DTO** is used in this article.
```

A DTO may be used to:

* Prevent over-posting.
* Hide properties that clients are not supposed to view.
* Omit some properties in order to reduce payload size.
* Flatten object graphs that contain nested objects. Flattened object graphs can be more convenient for clients.

> Update the `TodoItem` class to include a secret field:

Reviewer: ```suggestion
To demonstrate this here, let's update the `TodoItem` class to include a secret field:
```

[!code-csharp[](first-web-api/samples/3.0/TodoApiDTO/Models/TodoItem.cs?name=snippet&highlight=6)]

> The secret field needs to be hidden from this app, an administrative app could expose it.

Reviewer 2: ```suggestion
The secret field needs to be hidden from this app, but an administrative app could choose to expose it.
```

Verify you can post and get the secret field.

> Create an DTO model:

Reviewer 2: ```suggestion
Create a DTO model:
```

[!code-csharp[](first-web-api/samples/3.0/TodoApiDTO/Models/TodoItemDTO.cs?name=snippet)]

Update the `TodoItemsController` to use `TodoItemDTO`:

[!code-csharp[](first-web-api/samples/3.0/TodoApiDTO/Controllers/TodoItemsController.cs?name=snippet)]

> Verify you can't post and get the secret field.

Reviewer 2: ```suggestion
Verify you can't post or get the secret field.
```
